### PR TITLE
HTTP/2 negotiated connection pipeline config

### PR DESCRIPTION
**Pull request by rnro**
Motivation:

This continues the work to expose functionality which allows users to interact with HTTP/2 connections via async abstractions and using structured concurrency.

This work enables protocol negotiation between HTTP/1.1 and HTTP/2 in its most generic form.

Modifications:

Enable protocol negotiation between HTTP/1.1 and HTTP/2 using typed negotiation results (`NIOProtocolNegotiationResult`) from the `NIOTypedApplicationProtocolNegotiationHandler`.
In the HTTP/2 case the negotiation result will return the HTTP/2 connection channel and the `NIOHTTP2Handler.AsyncStreamMultiplexer` which exposes inbound streams as an iterable async stream.

Result:

Users will be able to set up negotiated HTTP/1.1 / HTTP/2 connections and iterate over inbound streams.

### CODE REVIEW

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 1]
```diff
@@ -380,6 +380,63 @@ extension Channel {
         return self.pipeline.addHandler(alpnHandler)
     }
 
+
+    /// Configures a channel to perform a HTTP/2 secure upgrade with typed negotiation results.
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:384]: NIT
```suggestion
    /// Configures a channel to perform an HTTP/2 secure upgrade with typed negotiation results.
```

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 2]
```diff
@@ -380,6 +380,63 @@ extension Channel {
...
+    ///         channel has been fully mutated.
+    ///   - http1ChannelConfigurator: A callback that will be invoked if HTTP/1.1 has been explicitly
+    ///         negotiated, or if no protocol was negotiated. Must return a future that completes when the
+    ///         channel has been fully mutated.
+    /// - Returns: An `EventLoopFuture` containing a ``NIOProtocolNegotiationResult`` that completes when the channel
+    ///     is ready to negotiate. The ``NIOProtocolNegotiationResult`` may itself be waited on to retrieve the result of
+    ///     the negotiation
+    @_spi(AsyncChannel)
+    public func configureHTTP2AsyncSecureUpgrade<HTTP1Output, HTTP2Output>(
+        h2ChannelConfigurator: @escaping (Channel) throws -> EventLoopFuture<HTTP2Output>,
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:413]: Can we call this `http2ChannelConfigurator` to align with the other parameter. I think these closures also need to be `@Sendable`.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 3]
```diff
@@ -725,3 +839,10 @@ extension ChannelPipeline.SynchronousOperations {
         }
     }
 }
+
+/// `NegotiationResult` is a generic negotiation result holder for HTTP/1.1 and HTTP/2
+@_spi(AsyncChannel)
+public enum NegotiationResult<HTTP1Output, HTTP2Output> {
+    case http1_1(HTTP1Output)
+    case http2(HTTP2Output)
+}
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:848]: I think we need to be a bit more specific here. Especially if we want to have something that does H1, H2, H3 in the future. What about naming this `HTTP1And2NegotiationResult`? Also open for better names cc @Lukasa @glbrntt

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 4]
```diff
@@ -460,40 +556,58 @@ extension Channel {
...
+    public func configureAsyncHTTPServerPipeline<HTTP1Output, HTTP2Output, HTTP2StreamOutput>(
         connectionConfiguration: NIOHTTP2Handler.ConnectionConfiguration,
         streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
         streamDelegate: NIOHTTP2StreamDelegate? = nil,
-        h2ConnectionChannelConfigurator: ((Channel) -> EventLoopFuture<Void>)? = nil,
-        configurator: @escaping NIOHTTP2Handler.StreamInitializer
-    ) -> EventLoopFuture<Void> {
-        return self._commonHTTPServerPipeline(configurator: configurator, h2ConnectionChannelConfigurator: h2ConnectionChannelConfigurator) { channel in
-            channel.configureHTTP2Pipeline(
+        h1ConnectionChannelConfigurator: @escaping NIOHTTP2Handler.ConnectionInitializerWithOutput<HTTP1Output>,
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:584]: This uses a type on the H2 handler which is a bit confusing.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 5]
```diff
@@ -460,40 +556,58 @@ extension Channel {
...
+    ///         These will be sent to the peer as part of the handshake.
+    ///   - streamConfiguration: The settings that will be used when establishing new HTTP/2 streams.
+    ///         These mainly pertain to flow control.
     ///   - streamDelegate: The delegate to be notified in the event of stream creation and close.
+    ///   - h1ConnectionChannelConfigurator: An optional callback that will be invoked only
+    ///         when the negotiated protocol is HTTP/1.1 to configure the connection channel.
     ///   - h2ConnectionChannelConfigurator: An optional callback that will be invoked only
-    ///         when the negotiated protocol is H2 to configure the connection channel.
+    ///         when the negotiated protocol is HTTP/2 to configure the connection channel.
     ///   - inboundStreamInitializer: A closure that will be called whenever the remote peer initiates a new stream.
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:574]: This doc comment doesn't match with the parameter names

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 6]
```diff
@@ -460,40 +556,58 @@ extension Channel {
...
         streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
         streamDelegate: NIOHTTP2StreamDelegate? = nil,
-        h2ConnectionChannelConfigurator: ((Channel) -> EventLoopFuture<Void>)? = nil,
-        configurator: @escaping NIOHTTP2Handler.StreamInitializer
-    ) -> EventLoopFuture<Void> {
-        return self._commonHTTPServerPipeline(configurator: configurator, h2ConnectionChannelConfigurator: h2ConnectionChannelConfigurator) { channel in
-            channel.configureHTTP2Pipeline(
+        h1ConnectionChannelConfigurator: @escaping NIOHTTP2Handler.ConnectionInitializerWithOutput<HTTP1Output>,
+        h2ConnectionChannelConfigurator: @escaping NIOHTTP2Handler.ConnectionInitializerWithOutput<HTTP2Output>,
+        streamInitializer: @escaping NIOHTTP2Handler.StreamInitializerWithOutput<HTTP2StreamOutput>
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:586]: I think we should holistically prefix the parameters with either h1/2 or http1/http2 if they are for one specific protocol. I got quite confused trying to understand what parameter applies to what

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 7]
```diff
@@ -460,40 +556,58 @@ extension Channel {
...
+    ) throws -> EventLoopFuture<NIOProtocolNegotiationResult<NegotiationResult<HTTP1Output, (HTTP2Output, NIOHTTP2Handler.AsyncStreamMultiplexer<HTTP2StreamOutput>)>>> {
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:587]: I am a bit surprised by this return type. I rather expected us to return the handler instead of the result here. The reason for this is that our bootstrap methods require the channel initialiser to return a `Handler` which we then use to await for the result.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 5 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:574]: I _think_ they do but that the diff is presenting them in a confusing way

Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift]: I restructured the original commit to split out the simple moving of a function to group like functions. Hopefully that makes the main diff easier to understand

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 8]
```diff
@@ -434,6 +491,47 @@ extension Channel {
         }
     }
 
+    /// Configures a `ChannelPipeline` to speak either HTTP or HTTP/2 according to what can be negotiated with the client.
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:494]:
```suggestion
    /// Configures a `ChannelPipeline` to speak either HTTP/1.1 or HTTP/2 according to what can be negotiated with the client.
```

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 9]
```diff
@@ -434,6 +491,47 @@ extension Channel {
...
+    ///         handshake.
+    ///   - streamConfiguration: The settings that will be used when establishing new HTTP/2 streams. These mainly pertain to flow control.
+    ///   - streamDelegate: The delegate to be notified in the event of stream creation and close.
+    ///   - h2ConnectionChannelConfigurator: An optional callback that will be invoked only
+    ///         when the negotiated protocol is H2 to configure the connection channel.
+    ///   - configurator: A callback that will be invoked after a protocol has been negotiated.
+    ///         The callback only needs to add application-specific handlers and must return a future
+    ///         that completes when the channel has been fully mutated.
+    /// - Returns: `EventLoopFuture<Void>` that completes when the channel is ready.
+    public func configureCommonHTTPServerPipeline(
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:514]: Why are we adding this? It doesn't seem related to the async work.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 6 above)
Comment (glbrntt) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:588]: Yup, agree, `streamInitializer` is not obviously http2

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 9 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:514]: 🤦 we aren't - it's just that git is having difficulty with the diff. Before my change we had three functions with `func configureCommonHTTPServerPipeline` and one `func _common` which two of the three called. I re-ordered the functions so that the three similarly named functions were all together and followed by the common code. This apparently baffles git even when it's split out into its own commit.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 3 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:850]: I've gone with `NIONegotiatedHTTPVersion`

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 4 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:586]: I've created a pair of simplified global types which are used everywhere except for where the existing `NIOHTTP2Handler.StreamInitializer` was already used.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 7 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:589]: I've changed this to match the bootstrap methods.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 10]
```diff
@@ -380,7 +392,60 @@ extension Channel {
...
+    ///         negotiated, or if no protocol was negotiated. Must return a future that completes when the
+    ///         channel has been fully mutated.
+    ///   - http2ChannelConfigurator: A callback that will be invoked if HTTP/2 has been negogiated, and that
+    ///         should configure the channel for HTTP/2 use. Must return a future that completes when the
+    ///         channel has been fully mutated.
+    /// - Returns: An `EventLoopFuture` containing a ``NIOTypedApplicationProtocolNegotiationHandler`` that completes when the channel
+    ///     is ready to negotiate. This can then be used to access the ``NIOProtocolNegotiationResult`` which may itself
+    ///     be waited on to retrieve the result of the negotiation
+    @_spi(AsyncChannel)
+    public func configureHTTP2AsyncSecureUpgrade<HTTP1Output, HTTP2Output>(
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:424]: Are we intending to make this public in the end? I don't see a lot of usage for this without adding the protocol handlers in there so maybe we should make it non public in the end and right now add an `_` to the method name?

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 11]
```diff
@@ -457,43 +563,62 @@ extension Channel {
...
+    ///   - http2StreamInitializer: A closure that will be called whenever the remote peer initiates a new stream.
+    /// - Returns: An `EventLoopFuture` containing a ``NIOTypedApplicationProtocolNegotiationHandler`` that completes when the channel
+    ///     is ready to negotiate. This can then be used to access the ``NIOProtocolNegotiationResult`` which may itself
+    ///     be waited on to retrieve the result of the negotiation
+    @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+    @_spi(AsyncChannel)
+    public func configureAsyncHTTPServerPipeline<HTTP1Output, HTTP2Output, HTTP2StreamOutput>(
         connectionConfiguration: NIOHTTP2Handler.ConnectionConfiguration,
         streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
         streamDelegate: NIOHTTP2StreamDelegate? = nil,
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:593]: IMO we should prefix those methods with `http2` as well. They only apply to it and right now you can only discover this by reading the docs.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 12]
```diff
@@ -457,43 +563,62 @@ extension Channel {
...
+    ///         when the negotiated protocol is HTTP/1.1 to configure the connection channel.
+    ///   - http2ChannelConfigurator: An optional callback that will be invoked only
+    ///         when the negotiated protocol is HTTP/2 to configure the connection channel.
+    ///   - http2StreamInitializer: A closure that will be called whenever the remote peer initiates a new stream.
+    /// - Returns: An `EventLoopFuture` containing a ``NIOTypedApplicationProtocolNegotiationHandler`` that completes when the channel
+    ///     is ready to negotiate. This can then be used to access the ``NIOProtocolNegotiationResult`` which may itself
+    ///     be waited on to retrieve the result of the negotiation
+    @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+    @_spi(AsyncChannel)
+    public func configureAsyncHTTPServerPipeline<HTTP1Output, HTTP2Output, HTTP2StreamOutput>(
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:590]: What about naming the `HTTP2Ouput` generic type as `HTTP2ConnectionOutput` to better differentiate the two H2 types. (Maybe also add it for the H1 type

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 13]
```diff
@@ -457,43 +563,62 @@ extension Channel {
...
         streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
         streamDelegate: NIOHTTP2StreamDelegate? = nil,
-        h2ConnectionChannelConfigurator: ((Channel) -> EventLoopFuture<Void>)? = nil,
-        configurator: @escaping NIOHTTP2Handler.StreamInitializer
-    ) -> EventLoopFuture<Void> {
-        return self._commonHTTPServerPipeline(configurator: configurator, h2ConnectionChannelConfigurator: h2ConnectionChannelConfigurator) { channel in
-            channel.configureHTTP2Pipeline(
+        http1ChannelConfigurator: @escaping NIOChannelInitializerWithOutput<HTTP1Output>,
+        http2ChannelConfigurator: @escaping NIOChannelInitializerWithOutput<HTTP2Output>,
+        http2StreamInitializer: @escaping NIOChannelInitializerWithOutput<HTTP2StreamOutput>
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:596]: Should we add `inbound` in this parameters name?

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 14]
```diff
@@ -291,6 +316,201 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
...
+        var receivedParts: [HTTPServerRequestPart] = []
+
+        func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+            self.receivedParts.append(self.unwrapInboundIn(data))
+            context.fireChannelRead(data)
+        }
+    }
+
+    /// A simple channel handler that records inbound frames.
+    class HTTP1ClientResponseRecorderHandler: ChannelInboundHandler {
```
Comment (FranzBusch) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:504]:
```suggestion
    final class HTTP1ServerRequestRecorderHandler: ChannelInboundHandler {
        typealias InboundIn = HTTPServerRequestPart

        var receivedParts: [HTTPServerRequestPart] = []

        func channelRead(context: ChannelHandlerContext, data: NIOAny) {
            self.receivedParts.append(self.unwrapInboundIn(data))
            context.fireChannelRead(data)
        }
    }

    /// A simple channel handler that records inbound frames.
    final class HTTP1ClientResponseRecorderHandler: ChannelInboundHandler {
```

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 15]
```diff
@@ -291,6 +316,201 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
...
+            // Let's try sending some http/1.1 requests.
+            // we need to put these through a mapping to remove references to `IOData` which isn't Sendable
+            try await self.clientChannel.writeOutbound(HTTP1ClientSendability.RequestPart.head(ConfiguringPipelineAsyncMultiplexerTests.requestHead))
+            try await self.clientChannel.writeAndFlush(HTTP1ClientSendability.RequestPart.end(nil)).get()
+            try await self.interactInMemory(self.clientChannel, self.serverChannel)
+        }
+
+        // check expectations
+        let clientRecorder = try await self.clientChannel.pipeline.handler(type: HTTP1ClientResponseRecorderHandler.self).get()
+        let serverRecorder = try await self.serverChannel.pipeline.handler(type: HTTP1ServerRequestRecorderHandler.self).get()
```
Comment (FranzBusch) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:453]: This is not safe. Those two types are not `Sendable`

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 11 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:593]: Configuration has been largely refactored.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 13 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift]: Configuration has been largely refactored.

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift (hunk 15 above)
Comment (rnro) [reply Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift]: I've made these recorder types `Sendable`

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 16]
```diff
@@ -434,6 +468,47 @@ extension Channel {
...
+    ///   - configurator: A callback that will be invoked after a protocol has been negotiated.
+    ///         The callback only needs to add application-specific handlers and must return a future
+    ///         that completes when the channel has been fully mutated.
+    /// - Returns: `EventLoopFuture<Void>` that completes when the channel is ready.
+    public func configureCommonHTTPServerPipeline(
+        connectionConfiguration: NIOHTTP2Handler.ConnectionConfiguration,
+        streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
+        streamDelegate: NIOHTTP2StreamDelegate? = nil,
+        h2ConnectionChannelConfigurator: ((Channel) -> EventLoopFuture<Void>)? = nil,
+        configurator: @escaping NIOChannelInitializer
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:496]: 's/configurator/initializer'

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 17]
```diff
@@ -434,6 +468,47 @@ extension Channel {
...
+    ///         when the negotiated protocol is H2 to configure the connection channel.
+    ///   - configurator: A callback that will be invoked after a protocol has been negotiated.
+    ///         The callback only needs to add application-specific handlers and must return a future
+    ///         that completes when the channel has been fully mutated.
+    /// - Returns: `EventLoopFuture<Void>` that completes when the channel is ready.
+    public func configureCommonHTTPServerPipeline(
+        connectionConfiguration: NIOHTTP2Handler.ConnectionConfiguration,
+        streamConfiguration: NIOHTTP2Handler.StreamConfiguration,
+        streamDelegate: NIOHTTP2StreamDelegate? = nil,
+        h2ConnectionChannelConfigurator: ((Channel) -> EventLoopFuture<Void>)? = nil,
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:495]: can this be `NIOChannelInitializer? = nil`

Comment (glbrntt) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:495]: also 's/h2/http2'

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 18]
```diff
@@ -457,43 +532,55 @@ extension Channel {
...
+    ///     is HTTP/1.1 to configure the connection channel.
+    ///   - http2ConnectionInitializer: An optional callback that will be invoked only when the negotiated protocol
+    ///     is HTTP/2 to configure the connection channel.
+    ///   - http2InboundStreamInitializer: A closure that will be called whenever the remote peer initiates a new stream.
+    /// - Returns: An `EventLoopFuture` containing a ``NIOTypedApplicationProtocolNegotiationHandler`` that completes when the channel
+    ///     is ready to negotiate. This can then be used to access the ``NIOProtocolNegotiationResult`` which may itself
+    ///     be waited on to retrieve the result of the negotiation
+    @available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+    @_spi(AsyncChannel)
+    public func configureAsyncHTTPServerPipeline<HTTP1ConnectionOutput, HTTP2ConnectionOutput, HTTP2StreamOutput>(
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:555]: I think the existing version of this has a notion of "upgrade" in its name

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 19]
```diff
@@ -673,55 +725,112 @@ extension ChannelPipeline.SynchronousOperations {
     /// 
     /// Using this rather than implementing a similar function yourself allows that pipeline to evolve without breaking your code.
     ///
+
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:728]:
```suggestion
```

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 20]
```diff
@@ -673,55 +725,112 @@ extension ChannelPipeline.SynchronousOperations {
...
+                    isOutboundHalfClosureEnabled: inboundStreamAsyncChannelConfiguration.isOutboundHalfClosureEnabled
                 )
             }
         }
     }
 }
+
+/// `NIONegotiatedHTTPVersion` is a generic negotiation result holder for HTTP/1.1 and HTTP/2
+@_spi(AsyncChannel)
+public enum NIONegotiatedHTTPVersion<HTTP1Output, HTTP2Output> {
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:770]: I think we require this:

```suggestion
public enum NIONegotiatedHTTPVersion<HTTP1Output: Sendable, HTTP2Output: Sendable>: Sendable {
```

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 21]
```diff
@@ -673,55 +725,112 @@ extension ChannelPipeline.SynchronousOperations {
...
+    case http1_1(HTTP1Output)
+    case http2(HTTP2Output)
+}
+
+/// `HTTP2AsyncConfiguration` contains all configuration required for setting up an HTTP/2 async connection.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+@_spi(AsyncChannel)
+public struct HTTP2AsyncConfiguration<HTTP2ConnectionInbound, HTTP2ConnectionOutbound, HTTP2StreamInbound, HTTP2StreamOutbound> {
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:778]:
```suggestion
public struct HTTP2AsyncConfiguration<HTTP2ConnectionInbound, HTTP2ConnectionOutbound, HTTP2StreamInbound, HTTP2StreamOutbound>: Sendable {
```

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 22]
```diff
@@ -673,55 +725,112 @@ extension ChannelPipeline.SynchronousOperations {
...
+        let inboundType: Inbound.Type
+        let outboundType: Outbound.Type
+        let backpressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark?
+        let isOutboundHalfClosureEnabled: Bool
+
+        public init(
+            inboundType: Inbound.Type = Inbound.self,
+            outboundType: Outbound.Type = Outbound.self,
+            backpressureStrategy: NIOAsyncSequenceProducerBackPressureStrategies.HighLowWatermark? = nil,
+            isOutboundHalfClosureEnabled: Bool? = nil
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:827]: I think these were added in a different order in NIO: `inboundType` and `outboundType` came at the end

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 23]
```diff
@@ -57,6 +60,23 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
         }
     }
 
+    final class HTTP1OKResponder: ChannelInboundHandler {
+        typealias InboundIn = HTTPServerRequestPart
+        typealias OutboundOut = HTTPServerResponsePart
+
+        func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+            guard case .end = self.unwrapInboundIn(data) else {
```
Comment (glbrntt) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:68]: nit: prefer switching over all cases

File: Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift [hunk 24]
```diff
@@ -218,77 +231,222 @@ final class ConfiguringPipelineAsyncMultiplexerTests: XCTestCase {
...
+            context.write(self.wrapOutboundOut(httpServerResponsePart), promise: promise)
+        }
+    }
+
+    /// A simple channel handler that records inbound messages.
+    internal final class InboundRecorderHandler<message>: ChannelInboundHandler, @unchecked Sendable {
+        typealias InboundIn = message
+
+        private let partsLock = NIOLock()
+        var _receivedParts: [message] = []
```
Comment (glbrntt) [inline Tests/NIOHTTP2Tests/ConfiguringPipelineAsyncMultiplexerTests.swift:432]:
```suggestion
        private var _receivedParts: [message] = []
```

File: Tests/NIOHTTP2Tests/TestUtilities.swift [hunk 25]
```diff
@@ -64,6 +64,8 @@ extension XCTestCase {
 
     /// Have two `NIOAsyncTestingChannel` objects send and receive data from each other until
     /// they make no forward progress.
+    ///
+    /// ** This function is not thread safe and can lead to deadlocks, prefer the one-way variant which is less error-prone**
```
Comment (glbrntt) [inline Tests/NIOHTTP2Tests/TestUtilities.swift:68]: It _is_ thread-safe, the problem is that calling it from multiple threads is racy.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 18 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:555]: The method which mentions "upgrade" is one layer down, the analogue of this is `configureCommonHTTPServerPipeline` - but we don't have any common configuration because we don't translate H2 into H1 so that's not relevant.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 16 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:496]: This code is existing API which has been moved so I can't rename it 😭

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift (hunk 17 above)
Comment (rnro) [reply Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:495]: This code is existing API which has been moved so I can't rename it 😭

**Comment by glbrntt** [comment]
oh, looks like you regressed allocations @rnro

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 26]
```diff
@@ -675,53 +727,109 @@ extension ChannelPipeline.SynchronousOperations {
...
+@_spi(AsyncChannel)
+public enum NIONegotiatedHTTPVersion<HTTP1Output: Sendable, HTTP2Output: Sendable> {
+    case http1_1(HTTP1Output)
+    case http2(HTTP2Output)
+}
+
+/// `HTTP2AsyncConfiguration` contains all configuration required for setting up an HTTP/2 async connection.
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+@_spi(AsyncChannel)
+public struct HTTP2AsyncConfiguration<HTTP2ConnectionInbound: Sendable, HTTP2ConnectionOutbound: Sendable, HTTP2StreamInbound: Sendable, HTTP2StreamOutbound: Sendable>: Sendable {
```
Comment (FranzBusch) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:777]:
```suggestion
public struct NIOHTTP2AsyncConfiguration<HTTP2ConnectionInbound: Sendable, HTTP2ConnectionOutbound: Sendable, HTTP2StreamInbound: Sendable, HTTP2StreamOutbound: Sendable>: Sendable {
```

**Comment by rnro** [comment]
@swift-server-bot test this please

**Comment by rnro** [comment]
Rebased on top of https://github.com/apple/swift-nio-http2/pull/409

**Comment by rnro** [comment]
All of the API breakage is in code added since the last release. Almost all of it is `Async...` configuration functions, there are also some closure type aliases which they used but we've simplified and done away with.

File: Sources/NIOHTTP2/HTTP2PipelineHelpers.swift [hunk 27]
```diff
@@ -677,55 +725,88 @@ extension ChannelPipeline.SynchronousOperations {
...
+        self.stream = stream
+        self.connectionInitializer = connectionInitializer
+        self.inboundStreamInitializer = inboundStreamInitializer
+    }
+}
+
+#if swift(>=5.7)
+@available(macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0, *)
+@_spi(AsyncChannel)
+extension NIOHTTP2AsyncConfiguration: @unchecked Sendable {}
```
Comment (glbrntt) [inline Sources/NIOHTTP2/HTTP2PipelineHelpers.swift:811]: nit: add a comment as to why this is `@unchecked`

**Review by glbrntt — APPROVED** [review]
